transfer/transfer.go

```diff
@@ -0,0 +1,177 @@
+// Server creates an rsync server on the destination
+type Server interface {
+	// Endpoint returns the endpoint used by the transfer
+	Endpoint() endpoint.Endpoint
+	// Transport returns the transport used by the transfer
+	Transport() transport.Transport
+	// ListenPort returns the port on which transfer server pod is listening on
+	ListenPort() int32
+	// IsHealthy returns whether or not all Kube resources used by endpoint are healthy
+	IsHealthy(ctx context.Context, c client.Client) (bool, error)
+	// Completed returns whether or not the current attempt of transfer is completed
+	Completed(ctx context.Context, c client.Client) (bool, error)
+	// PVCs returns the list of PVCs the transfer will migrate
+	PVCs() []*corev1.PersistentVolumeClaim
+	// MarkForCleanup add the required labels to all the resources for
+	// cleaning up
+	MarkForCleanup(ctx context.Context, c client.Client, key, value string) error
+}
+
+type Client interface {
+	// Transport returns the transport used by the transfer
+	Transport() transport.Transport
+	// PVCs returns the list of PVCs the transfer will migrate
+	PVCs() []*corev1.PersistentVolumeClaim
+	// IsCompleted returns whether the client is done
+	Status(ctx context.Context, c client.Client) (*Status, error)
+	// MarkForCleanup adds a key-value label to all the resources to be cleaned up
+	MarkForCleanup(ctx context.Context, c client.Client, key, value string) error
+}
+
+// PodOptions allow callers to pass custom configuration for the transfer pods
+type PodOptions struct {
+	// PodSecurityContext determines what GID the rsync process gets
+	// In case of shared storage SupplementalGroups is configured to get the gid
+	// In case of block storage FSGroup is configured to get the gid
+	PodSecurityContext corev1.PodSecurityContext
+	// ContainerSecurityContext determines what selinux labels, UID and drop capabilities
+	// are required for the containers in rsync transfer pod via SELinuxOptions, RunAsUser and
+	// Capabilities fields respectively
+	ContainerSecurityContext corev1.SecurityContext
+	// NodeName allows pods to be scheduled on a specific node. This is especially required in
+	// client pods where the PVC's are bound to a specific regions and the node where the pod runs on
+	// has to be in the same region as the PVC to be scheduled and bound.
+	NodeName string
+	// NodeSelector is a wider net for scheduling the pods on node than NodeName.
+	NodeSelector map[string]string
+	// Resources allows for configuring the resources consumed by the transfer pods. In general
+	// it is good to provision destination transfer pod with same or larger resources than the source
+	// so that the network is not congested.
+	Resources corev1.ResourceRequirements
+}
+
+type Status struct {
+	Running   *Running
+	Completed *Completed
+}
+
+type Running struct {
+	StartedAt *metav1.Time
+}
+
+type Completed struct {
+	Successful bool
+	Failure    bool
+	FinishedAt *metav1.Time
+}
```

Comment:
1. Running and Completed as sub-structs because it would allow for adding more information to the status in the future like Pending (if an rsync pod is stuck at trying to mount to a PVC, for instance). Reporting progress could be another example that could go in the Running struct.

2. Successful and Failure both because of the likeliness of a third event of unknown. Although I should convert it to both of them to *bool to avoid defaulting it to false.

I would be curious to hear if there are alternatives to consider here.